### ISSUE TITLE
https websites not blocked immediately

### ISSUE THREAD

**Issue by Dalacor** [e2guardian/e2guardian#594]
I am not sure if this is a E2guardian bug or not, but I hope that this can be addressed.

I have noticed that if you block say gambling category and take a website like www.888.casino.com. When you click on a link to this website, the Internet Router goes to the Internet doing a DNS query and comes back and then the website is blocked as https://www.888casino.com

However, if you type in www.888casino.com in the address bar, the website is blocked immediately without going to the Internet and the page shows as www.888casino.com not as https://www.888casino.com.

If you do the same thing for say www.games.co.uk - regardless of whether you type it in the address bar or click on a link in google search, the website is only blocked after the Internet Router does a DNS query and both methods display the page as https://www.games.co.uk

If the blacklist contains games.co.uk - I can't see why this website is connecting to the Internet first regardless of whether you type in the address or click on a link in google search. Why do https websites need to connect to the Internet first.

In the normal way this does not cause a problem if you are using MITM as the access denied page that you get back is identical.

The reason why this is a problem for me is because I have three blacklists included in E2guardian - The Original Big Blacklist (urlblacklist.com), but then they closed, so I bought a subscription to Squidblacklist.org (but the owner passed away and the blacklists are no longer updated) and now I am trying UT1 Blacklist which is quite good. However despite having three blacklists, there are gaps in the filtering especially with gambling websites and I am particularly concerned about malware protection now.

I looked around for any blacklists that I could subscribe to, but all the ones that I found seem to be self maintained nowadays which means that they simply don't have the ability to have a comprehensive blacklist of all websites especially malware ones.

So I bought a Blacklist subscription from Draytek - Cyren for all my Draytek Routers and this has helped to fill in the gaps as I can see virtually every gambling website is blocked and their malware protection (Cyren) is second to none!

But the problem is that if the website is blocked by both E2guardian blacklists and Draytek Cyren blacklists then instead of getting the message "blocked site 888casino.com" when clicking on the link in Google Search, I am instead getting "Failed to nogotiate ssl connection to server" because what is happening is that the router is checking the website and then blocking it (through the Draytek blacklist) and the client is getting that error message because the request is dropped by Draytek Router (not by E2guardian).

Due to the fact that it has become virtually impossible to find a downloadable blacklist that has sufficient protection anymore (especially against malware) as every downloadable blacklist that I looked at is largely a group of people manually adding websites to the blacklist, it has become clear that the only option is to subscribe to a blacklist service like Cyren or Webroot etc. However you cannot download their blacklists - you can only subscribe to it through your Internet routers.

Therefore I think that E2guardian needs to support both downloaded blacklists within E2guardian and blacklists that are maintained on the Internet router such that E2guardian shows the same message i.e. - Website name blocked - not "Failed to nogotiate ssl connection to server"

This can be done in two steps hopefully.

One get E2guardian to block the https website before it connects to the Internet, so that it always uses E2guardian's internal blacklists first before using the Draytek blacklist.

Second, for instances when the website is not in any of the E2guardian blacklists ,but is in the Draytek blacklists, then E2guardian needs to accept the access denied message (or whatever it is receiving from Draytek) and show website blocked, not failed to negotiate ssl connection. I suspect step two is dependent on step one!

E2guardian is brilliant because of the sheer level of customisabilty with groups, filetypes blocking, google search terms, safe search, word content filtering etc! But it is no longer feasible to rely on downloaded blacklists anymore as there are no businesses providing a downloadable blacklist subscription service that is sufficiently comprehensive anymore - especially as malware protection has become absolutely critical nowadays.

Therefore the way the E2guardian blacklist checking currently works will need to change in the future to accommodate the way commercial blacklists work nowadays, where you subscribe through the Internet router.

Thank you.

**Comment by philipianpearce** [comment]
@Dalacor. Can you provide a few more details? - 

What version of e2g are you using? 

Platform? (Linux or BSD) 

For the cases you describe are you using explicit proxy or transparent on the user device? What browser? 

Do you have the router defined as the upstream proxy in e2g or is e2g configured to go direct? 

If you leave e2g out of the path and just have the router what message do you get for the two different urls? Please post the entire html response, including headers. What is the case with explicit proxy? Transparent, if used? Do you have to install a root CA cert on the browser, in order to get a sensible response? 

Philip 

----- Original Message -----

**Comment by Dalacor** [comment]
@philipianpearce 

Running E2g 5.3.3, Linux Slackware Current

I have tried both transparent and explicit. With explicit I get the error messages shown above. With transparent I get the usual error message that the browser gives when E2g blocks an https website and MITM is not enabled. I use Firefox, but I have tested Chrome as well with similar results.

I tried your suggestion of setting up router as upstream proxy but could not get the Internet to work at all for any website, whether I used port 80 or port 443 for the upstream proxy (router). The router is not a proxy server in that sense, so I can understand why the Internet won't work in that scenario.

Yes I have the certificates installed and I only get the E2guardian template message with the certificates installed and MITM enabled for explicit proxy. Transparent proxy (is not using MITM) so I get the normal E2g error message about SSL error.

Without e2guardian I get this message when clicking on www.888casino.com link in Google Search

 "Warning: Potential Security Risk Ahead

Firefox detected a potential security threat and did not continue to www.888casino.com. If you visit this site, attackers could try to steal information like your passwords, emails, or credit card details."

If I click on advanced and accept risk and continue - I get this error message

<html><head><title>Black List </title></head><body><center><br><br><br><p>The requested Web page <br> from 192.168.1.10 <br>to www.888casino.com/ <br>that is categorized with [Black List] <br>has been blocked by DrayTek DNS Filter.<p>Please contact your system administrator for further information.</center></body><BR><center>[Powered by DrayTek]</center></html>

If I type in www.888casino.com in the address bar it goes straight to the Draytek error message.

Which is essentially what happens under E2guardian. If I put www.888casino.com it goes straight to the E2guardian template error message that I am expecting to see.. But when you click on the link and it redirects from http to https it is giving me that "Failed to nogotiate ssl connection to server" because the Draytek router is actually returning the "warning Potential Security Risk ahead" (in Firefox) - Chrome provides a similar but different message

"This site can’t be reached The connection was reset.
Try:

Checking the connection
Checking the proxy and the firewall
Running Windows Network Diagnostics
ERR_CONNECTION_RESET"

Basically what is happening is that when it is converting from http to https - it goes to the Internet first and then comes back. If it is only blocked by E2guardian, then E2guardian gets the website back (as it is expecting it) and then returns the E2g error message back to the client. But if the website is in the Draytek blacklist, then Draytek is returning the browser warning message that the website is insecure and compromised etc or the certificate is not secure, which is not what E2g is expecting to receive and thus it throws the ssl connection error.

Hope that helps explains it. Thank you.

**Comment by philipianpearce** [comment]
@Dalacor, See comments below. 

i will answer your other queries later, one we have cleared these ones up. 

Regards 
Philip 

----- Original Message -----

> From: "Dalacor" <notifications@github.com>
> To: "e2guardian/e2guardian" <e2guardian@noreply.github.com>
> Cc: "Philip Pearce" <philip.pearce@e2bn.org>, "Mention"
> <mention@noreply.github.com>
> Sent: Tuesday, 7 April, 2020 3:33:45 PM
> Subject: Re: [e2guardian/e2guardian] https websites not blocked
> immediately (#594)

> @philipianpearce
> Running E2g 5.3.3, Linux Slackware Current
> I have tried both transparent and explicit. With explicit I get the
> error messages shown above. With transparent I get the usual error
> message that the browser gives when E2g blocks an https website and
> MITM is not enabled. I use Firefox, but I have tested Chrome as well
> with similar results.
> I tried your suggestion of setting up router as upstream proxy but
> could not get the Internet to work at all for any website, whether I
> used port 80 or port 443 for the upstream proxy (router). The router
> is not a proxy server in that sense, so I can understand why the
> Internet won't work in that scenario.
I did not expect that to work, just checking how you had it set up .
> Yes I have the certificates installed and I only get the E2guardian
> template message with the certificates installed and MITM enabled
> for explicit proxy. Transparent proxy (is not using MITM) so I get
> the normal E2g error message about SSL error.
> Without e2guardian I get this message when clicking on
> www.888casino.com link in Google Search
> "Warning: Potential Security Risk Ahead
> Firefox detected a potential security threat and did not continue to
> www.888casino.com . If you visit this site, attackers could try to
> steal information like your passwords, emails, or credit card
> details."

 OK, the above is a Firefox message due to the router sending a faked
 server cert. (so this is not http, it must be https to get this
 message) - Once you have allowed the fake cert in a browser session
 the browser will remember that and not give the message on later
 requests to that site.

 You would need to load the Draytex generated CA root cert on browser
 in order to avoid this.

 (The Draytex system is a DNS web filter, so for blocks it probably
 returns via DNS the router's internal ip and then generates the
 block page on the router when the browser connects to that ip. Can
 you check this using dig or nslookup?)

> If I click on advanced and accept risk and continue - I get this
> error message <title>Black List </title>

> The requested Web page
> from 192.168.1.10
> to www.888casino.com/
> that is categorized with [Black List]
> has been blocked by DrayTek DNS Filter.
> Please contact your system administrator for further information.
> [Powered by DrayTek]
> If I type in www.888casino.com in the address bar it goes straight to
> the Draytek error message.

See above

> Which is essentially what happens under E2guardian. If I put
> www.888casino.com it goes straight to the E2guardian template error
> message that I am expecting to see.. But when you click on the link
> and it redirects from http to https it is giving me that "Failed to
> nogotiate ssl connection to server" because the Draytek router is
> actually returning the "warning Potential Security Risk ahead" (in
> Firefox) - Chrome provides a similar but different message
> "This site can’t be reached The connection was reset
> ERR_CONNECTION_RESET"
> Basically what is happening is that when it is converting from http
> to https - it goes to the Internet first and then comes back. If it
> is only blocked by E2guardian, then E2guardian gets the website back
> (as it is expecting it) and then returns the E2g error message back
> to the client. But if the website is in the Draytek blacklist, then
> Draytek is returning the browser warning message that the website is
> insecure and compromised etc or the certificate is not secure, which
> is not what E2g is expecting to receive and thus it throws the ssl
> connection error.

The "Failed to nogotiate ssl connection to server" message means that
 e2g is not able to do an SSLconnect to the DrayTex router, so with
 https it is NOT getting the Draytex error to pass back to the
 browser. 

This may be  a miss-match of tls protocols. To diagnose this switch on 'logsslerrors' in
 e2guardian.conf. Output will be in syslog.

 Also your e2g should be compiled/linked to OpenSSL v1.1 or later.
 There are a number of issues with previous versions.

**Comment by Dalacor** [comment]
Yes I am running openssl 1.1 because as you know, previous versions were causing problems with MITM.

Yes I was wondering if there was some certificate on the Draytek router that needs to be either imported in e2g or into the browsers on each computer. I will investigate this on Wednesday or Thursday when I have time.

I am wondering if perhaps you want to think about a subscription service for blacklists (especially malware) similar to how Draytek do it but based on your Protex system as downloadable blacklists are simply not adequate in today's world as there are no commercial companies doing this anymore. The Draytek Cyren way is now how "blacklists" are being managed at commercial level nowadays. You could do something similar for both blacklists and phraselists and basically have it baked into E2guardian.

**Comment by philipianpearce** [comment]
Note, I have edited my last post.  The e2g to draytex SSLconnect is failing before the server cert is checked so that needs to be sorted first before the draytex CA root issue.

Protex are currently researching into revised list generation so once that is complete, we will look again to see if if is feasible for us to make a subscription list service available.  I have had some ideas in the past as to how it might be implemented, but other areas of development have been more urgent. Such as the V5 re-writing and bug-fixing and the changes needed to work towards h2 support.

**Comment by Dalacor** [comment]
I would highly recommend it - going with some kind of subscription service. E2guardian is fantastic and the program itself works so well, but the lack of any commercial blacklists and phraselists means that even for home use, it is simply not as effective as it should be. Admittedly I am more concerned about malware protection these days.

I don't know if Protex use blacklists in the same way as E2guardian does, but just so that you know, my understanding is that Webroot/Cyren etc don't use blacklists so much as categorisation on demand I believe and it does seem to be the future as there are just far too many websites today for a blacklist to be effective anymore. 123movies and frivgames for example keep changing their url's all the time.

I will get back to you when I have investigated the SSL connect to Draytek issue as per your instructions. Thank you.

**Comment by Dalacor** [comment]
I have enabled logsslerrors - I actually had to add that setting to my file as it was not there!

The two error messages that I am getting are
"E2guardian (1157): ssl_connect failed to www.games.co.uk"
and
I am getting a lot of "E2guardian (1157): SSL _write failed - error"

The second error may not be related to the first.

I think that I am going to have to accept that I will get this "Failed to nogotiate ssl connection to server" error as I cannot see what I need to do to fix this issue. I have looked at both the router and E2guardian and I can't find anything in the logs of either to indicate why the SSL Connection is failing. However, it is blocking any website in the blacklist of either e2guardian or draytek and this is actually the most important thing.

I think it would make more sense to look ahead to the future and see about a Protex or other subscription service that is baked into E2guardian because the issue isn't really Draytek and Cyren per se. The underlying problem is that downloadable blacklists are not commercially available anymore and the user supported ones left are inadequate unfortunately so E2guardian blacklists need to move in a new direction.

I will close this question as I don't think it's warranted to spend time troubleshooting this unless you intend E2guardian to support Router enabled blacklists in the future. It is blocking the websites and that is actually the most important thing. Thank you.

**Comment by philipianpearce** [comment]
To pick up on an original point you made - in order to get the block page for ssl sites blocked by e2g v5.3, before a connect is attempted, override the library function sslrequestcheck  by adding the following in site.story:-

\# SSL request check - override built-in function
\#  returns true if exception, blocked or gomitm
function(sslrequestcheck)
if(true) returnif sslexceptioncheck
if(sitein, banned) setblock
if(true) returnif sslcheckmitm
if(blockset) return true
if(true) sslreplace
ifnot(returnset) returnif sslcheckblanketblock
if(true) setgrey

Note that this will also have the effect of disabling exception/greyurllists at blocked https sites, although if these sites are also blocked by draytek this is irrelevant.

v5.4 is now almost ready for pre-release.  A new feature is useoriginal ip which means that we have the destination IP at earlier stage than before in transparent mode.

I have now started work on v5.5.dev and intend to implement an earlier DNS look-up for explicit proxy.   This means that is will be possible to make storyboard decisions before the connection is made, based on the destination IP for both explicit and transparent proxy modes.   

So if an upstream DNS based filter system returns a consistent IP (or range) for blocks, then either a specific "upstream block" message can be returned to the user, or the request can be made an exception so that the upstream block page is displayed.  If you can confirm using dig that this is the case that would be helpful.  Having got that implemented then we can then consider whether a more closer integration is feasible or desirable as a standard feature.

**Comment by Dalacor** [comment]
I will try out your suggestions in the storyboard and see if that works now. I have uploaded the results of dig and nslookup - so you can see exactly what I have done as I am not entirely sure what you are looking for here. The draytek router is 10.12.8.4 by the way!

![nslookup](https://user-images.githubusercontent.com/6070032/79220948-dab30100-7e4c-11ea-97a3-03914e694716.png)

![dig](https://user-images.githubusercontent.com/6070032/79220954-dedf1e80-7e4c-11ea-87df-2bc9089ce989.png)

Ultimately I am happy to wait until E2guardian has an integrated solution so that you can support one method of accessing subscription lists because the websites are being blocked and that is in the end the most important concern. I will try the storyboard suggestion now.

**Comment by Dalacor** [comment]
WOW - that has helped a lot thank you! If I have understood this correctly, what this change does is that E2guardian checks the local blacklists and returns the usual access denied message before actually connecting to the Internet. Which is basically what I originally requested and perhaps should be implemented for all new releases unless there is some good reason why E2guardian was designed to connect to the Internet first before returning access denied?

It will mean that it is relatively painless to integrate E2guardian with any subscription based blacklist using the coding above presuming that one still uses the locally downloaded blacklists as well which I intend to do as it makes sense to use local blacklists first. Only websites that are not in E2guardian blacklists are now being affected with the SSL message. Obviously this won't solve the problem for a subscription only based blacklist setup.

I am now getting two different error messages. If the website is in the local E2guardian blacklist, it is returning the expected Access Denied Message that I have setup in E2guardian. If the website is not blacklisted in e2guardian, but is in the Draytek Blacklist, then I am getting the "Failed to nogotiate ssl connection to server" error message. However that is not a problem as 98% of the access denied messages are what we are expecting to see as the local blacklists do have the vast majority. I do believe that the Router is the cause of the problem, but I am unable to determine why.

I tested websites that are whitelisted in e2guardian (but still in the E2guardian blacklist) and I tested websites that I have explicitly blacklisted in E2guardian and it would appear that everything is working as it should be. I presume that I will only have an issue if I have whitelisted a website in E2guardian that is blacklisted by Draytek - in which case all I need to do is whitelist it in Draytek itself!

Thank you. I was not expecting to be able to fix this issue without big changes to E2g. Brilliant.
